Bug: https://trac.sagemath.org/ticket/23696

From 3c5d53c1f537acf545a0c6579feef1cff7cd541a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Mon, 15 Jan 2018 12:57:34 +1300
Subject: Rebase MPL-2.1.0 upgrade branch

---
 build/pkgs/backports_functools_lru_cache/SPKG.txt  |  5 ++
 .../backports_functools_lru_cache/checksums.ini    |  4 ++
 .../backports_functools_lru_cache/dependencies     |  5 ++
 .../package-version.txt                            |  1 +
 build/pkgs/backports_functools_lru_cache/type      |  1 +
 build/pkgs/matplotlib/checksums.ini                |  6 +--
 build/pkgs/matplotlib/dependencies                 |  2 +-
 build/pkgs/matplotlib/package-version.txt          |  2 +-
 build/pkgs/matplotlib/spkg-src                     |  6 +--
 build/pkgs/setuptools_scm/checksums.ini            |  6 +--
 build/pkgs/setuptools_scm/package-version.txt      |  2 +-
 build/pkgs/subprocess32/SPKG.txt                   |  5 ++
 build/pkgs/subprocess32/checksums.ini              |  4 ++
 build/pkgs/subprocess32/dependencies               |  5 ++
 build/pkgs/subprocess32/package-version.txt        |  1 +
 build/pkgs/subprocess32/spkg-install               |  3 ++
 build/pkgs/subprocess32/type                       |  1 +
 .../introduction.rst                               |  2 +-
 src/sage/all.py                                    |  2 -
 src/sage/plot/arc.py                               | 18 +++++--
 src/sage/plot/arrow.py                             | 12 ++---
 src/sage/plot/contour_plot.py                      | 12 ++---
 src/sage/plot/plot.py                              | 12 +++--
 src/sage/plot/plot3d/list_plot3d.py                | 55 ++++++++++------------
 src/sage/probability/probability_distribution.pyx  |  2 +-
 26 files changed, 111 insertions(+), 66 deletions(-)
 create mode 100644 build/pkgs/backports_functools_lru_cache/SPKG.txt
 create mode 100644 build/pkgs/backports_functools_lru_cache/checksums.ini
 create mode 100644 build/pkgs/backports_functools_lru_cache/dependencies
 create mode 100644 build/pkgs/backports_functools_lru_cache/package-version.txt
 create mode 100644 build/pkgs/backports_functools_lru_cache/spkg-install
 create mode 100644 build/pkgs/backports_functools_lru_cache/type
 create mode 100644 build/pkgs/subprocess32/SPKG.txt
 create mode 100644 build/pkgs/subprocess32/checksums.ini
 create mode 100644 build/pkgs/subprocess32/dependencies
 create mode 100644 build/pkgs/subprocess32/package-version.txt
 create mode 100644 build/pkgs/subprocess32/spkg-install
 create mode 100644 build/pkgs/subprocess32/type

From d109b7593ef7d1e8d3e234d554524e99c7e92b5e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Mon, 15 Jan 2018 15:44:38 +1300
Subject: WIP on inserting matplotlib style sheet as a plot option and
 defaulting it to classic

---
 src/sage/plot/graphics.py | 6 ++++++
 src/sage/plot/plot.py     | 6 ++++++
 2 files changed, 12 insertions(+)

From 9b90efadd3a8aa3dfc2ecd77c3f8e538f20cf424 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Tue, 16 Jan 2018 10:30:02 +1300
Subject: Set matplotlib options in sphinx_plot so that 3D plots look nice
 again.

---
 src/doc/common/conf.py | 3 +++
 1 file changed, 3 insertions(+)

From 4005ec3dae5ab697d215d084d1e7bddea7ff110e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Tue, 16 Jan 2018 14:48:26 +1300
Subject: Fix the arrow matching pattern

---
 src/sage/plot/arrow.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

From 3a6792ea300002114fc701acfb14ea52f72fe040 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Tue, 16 Jan 2018 23:18:59 +1300
Subject: Setting more parameters in sphinx_plot for better 3D documentation

---
 src/doc/common/conf.py | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

From e0534a8c68bc60cf2495acf32598096d64579184 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Mon, 22 Jan 2018 11:55:18 +1300
Subject: Make all doctests set agg as the backend for matplotlib.

---
 src/sage/doctest/control.py | 5 +++++
 1 file changed, 5 insertions(+)

From d5be931f48b77bb03d2d5750bea3805435e795fd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Thu, 25 Jan 2018 14:39:43 +1300
Subject: Change the startup test for numpy and pyparsing now that they are
 imported in the doctesting framework via matplotlib.

---
 src/sage/tests/startup.py | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)


--- a/sage/src/doc/en/thematic_tutorials/explicit_methods_in_number_theory/introduction.rst
+++ b/sage/src/doc/en/thematic_tutorials/explicit_methods_in_number_theory/introduction.rst
@@ -74,7 +74,7 @@
     sage: import warnings
     sage: warnings.simplefilter('ignore', UserWarning)
     sage: v = [[len(factor(n*m)) for n in [1..15]] for m in [1..15]]
-    sage: list_plot3d(v, interpolation_type='nn')
+    sage: list_plot3d(v, interpolation_type='clough')
     Graphics3d Object
 
 
--- a/sage/src/sage/all.py
+++ b/sage/src/sage/all.py
@@ -313,8 +313,6 @@
 # Ignore all deprecations from IPython etc.
 warnings.filterwarnings('ignore',
     module='.*(IPython|ipykernel|jupyter_client|jupyter_core|nbformat|notebook|ipywidgets|storemagic)')
-# Ignore warnings due to matplotlib-1.5 together with numpy-1.13
-warnings.filterwarnings('ignore', module='matplotlib[.]contour|numpy[.]ma[.]core')
 # However, be sure to keep OUR deprecation warnings
 warnings.filterwarnings('default',
     '[\s\S]*See http://trac.sagemath.org/[0-9]* for details.')
--- a/sage/src/sage/plot/arc.py
+++ b/sage/src/sage/plot/arc.py
@@ -297,16 +297,28 @@
             sage: a = arc((0,0),2,1,0,(pi/5,pi/2+pi/12), linestyle="--", color="red")
             sage: b = a[0].bezier_path()
             sage: b[0]
-            Bezier path from (1.618..., 0.5877...) to (-0.5176..., 0.9659...)
+            Bezier path from (1.133..., 0.8237...) to (-0.2655..., 0.9911...)
         """
         from sage.plot.bezier_path import BezierPath
         from sage.plot.graphics import Graphics
+        from matplotlib.path import Path
+        import numpy as np
         ma = self._matplotlib_arc()
-        transform = ma.get_transform().get_matrix()
+        def theta_stretch(theta, scale):
+            theta = np.deg2rad(theta)
+            x = np.cos(theta)
+            y = np.sin(theta)
+            return np.rad2deg(np.arctan2(scale * y, x))
+        theta1 = theta_stretch(ma.theta1, ma.width / ma.height)
+        theta2 = theta_stretch(ma.theta2, ma.width / ma.height)
+
+        pa = ma
+        pa._path = Path.arc(theta1, theta2)
+        transform = pa.get_transform().get_matrix()
         cA, cC, cE = transform[0]
         cB, cD, cF = transform[1]
         points = []
-        for u in ma._path.vertices:
+        for u in pa._path.vertices:
             x, y = list(u)
             points += [(cA * x + cC * y + cE, cB * x + cD * y + cF)]
         cutlist = [points[0: 4]]
--- a/sage/src/sage/plot/arrow.py
+++ b/sage/src/sage/plot/arrow.py
@@ -147,8 +147,8 @@
         bpath = Path(self.vertices, self.codes)
         p = FancyArrowPatch(path=bpath,
                             lw=width, arrowstyle='%s,head_width=%s,head_length=%s' % (style, head_width, head_length),
-                            fc=color, ec=color)
-        p.set_linestyle(get_matplotlib_linestyle(options['linestyle'], return_type='long'))
+                            fc=color, ec=color, 
+                            linestyle=get_matplotlib_linestyle(options['linestyle'], return_type='long'))
         p.set_zorder(options['zorder'])
         p.set_label(options['legend_label'])
         subplot.add_patch(p)
@@ -336,7 +336,7 @@
         this into account. See :trac:`12836`::
 
             sage: fig = Graphics().matplotlib()
-            sage: sp = fig.add_subplot(1,1,1)
+            sage: sp = fig.add_subplot(1,1,1, label='axis1')
             sage: a = arrow((0,0), (1,1))
             sage: b = arrow((0,0), (1,1), width=20)
             sage: p1 = a[0]._render_on_subplot(sp)
@@ -358,7 +358,7 @@
             sage: a.save(filename=filename)
             sage: with open(filename, 'r') as f:
             ....:     contents = f.read().replace('\n', ' ')
-            sage: two_stroke_pattern = r'setdash.*stroke.*stroke.*setdash'
+            sage: two_stroke_pattern = r'setdash.*stroke.*stroke.*setdash.*setdash'
             sage: import re
             sage: two_stroke_re = re.compile(two_stroke_pattern)
             sage: two_stroke_re.search(contents) is None
@@ -383,8 +383,8 @@
                             lw=width,
                             arrowstyle='%s,head_width=%s,head_length=%s' % (style, head_width, head_length),
                             shrinkA=arrowshorten_end, shrinkB=arrowshorten_end,
-                            fc=color, ec=color)
-        p.set_linestyle(get_matplotlib_linestyle(options['linestyle'], return_type='long'))
+                            fc=color, ec=color,
+                            linestyle=get_matplotlib_linestyle(options['linestyle'], return_type='long'))
         p.set_zorder(options['zorder'])
         p.set_label(options['legend_label'])
 
@@ -437,7 +437,7 @@
                             pe1.draw_path(renderer, gc, tpath, affine, rgbFace)
 
             pe1 = ConditionalStroke(CheckNthSubPath(p, 0), [pe.Stroke()])
-            pe2 = ConditionalStroke(CheckNthSubPath(p, 1), [pe.Stroke(linestyle="solid")])
+            pe2 = ConditionalStroke(CheckNthSubPath(p, 1), [pe.Stroke(dashes={'dash_offset': 0, 'dash_list': None})])
             p.set_path_effects([pe1, pe2])
 
         subplot.add_patch(p)
--- a/sage/src/sage/plot/contour_plot.py
+++ b/sage/src/sage/plot/contour_plot.py
@@ -178,12 +178,10 @@
         if fill:
             if contours is None:
                 CSF = subplot.contourf(self.xy_data_array, cmap=cmap,
-                                       extent=(x0, x1, y0, y1),
-                                       label=options['legend_label'])
+                                       extent=(x0, x1, y0, y1))
             else:
                 CSF = subplot.contourf(self.xy_data_array, contours, cmap=cmap,
-                                       extent=(x0, x1, y0, y1), extend='both',
-                                       label=options['legend_label'])
+                                       extent=(x0, x1, y0, y1), extend='both')
 
         linewidths = options.get('linewidths', None)
         if isinstance(linewidths, (int, Integer)):
@@ -200,13 +198,11 @@
         if contours is None:
             CS = subplot.contour(self.xy_data_array, cmap=cmap,
                                  extent=(x0, x1, y0, y1),
-                                 linewidths=linewidths, linestyles=linestyles,
-                                 label=options['legend_label'])
+                                 linewidths=linewidths, linestyles=linestyles)
         else:
             CS = subplot.contour(self.xy_data_array, contours, cmap=cmap,
                                  extent=(x0, x1, y0, y1),
-                                 linewidths=linewidths, linestyles=linestyles,
-                                 label=options['legend_label'])
+                                 linewidths=linewidths, linestyles=linestyles)
         if options.get('labels', False):
             label_options = options['label_options']
             label_options['fontsize'] = int(label_options['fontsize'])
--- a/sage/src/sage/plot/plot.py
+++ b/sage/src/sage/plot/plot.py
@@ -478,7 +478,7 @@
 
 ::
 
-    sage: plt.imshow([[(0,0,0)]])
+    sage: plt.imshow([[(0.0,0.0,0.0)]])
     <matplotlib.image.AxesImage object at ...>
     sage: plt.savefig(os.path.join(SAGE_TMP, 'foo.png'))
 
@@ -1011,6 +1011,12 @@
 
     - ``fillalpha`` - (default: 0.5) How transparent the fill is.
       A number between 0 and 1.
+      
+    MATPLOTLIB STYLE SHEET OPTION:
+    
+    - ``stylesheet`` - (Default: classic) Support for loading a full matplotlib style sheet.
+      Any style sheet listed in ``matplotlib.pyplot.style.available`` is acceptable. If a
+      non-existing style is provided the default classic is applied.
 
     EXAMPLES:
 
@@ -2908,11 +2914,14 @@
         If ``plotjoined`` is ``False`` then the axis that is in log scale
         must have all points strictly positive. For instance, the following
         plot will show no points in the figure since the points in the
-        horizontal axis starts from `(0,1)`.
+        horizontal axis starts from `(0,1)`. Further, matplotlib will display
+        a user warning.
 
         ::
 
             sage: list_plot(yl, scale='loglog')         # both axes are log
+            doctest:warning
+            ...
             Graphics object consisting of 1 graphics primitive
 
         Instead this will work. We drop the point `(0,1)`.::
@@ -3288,12 +3297,15 @@
 
         If ``plotjoined`` is ``False`` then the vertical axis must have all
         points strictly positive. Otherwise the plot will come up empty.
-        For instance the following plot contains a point at `(1,0)`.
+        For instance the following plot contains a point at `(1,0)`. Further, 
+        matplotlib will display a user warning.
 
         ::
 
             sage: xl = [2**k for k in range(12)]; yl = range(len(xl))
             sage: list_plot_semilogy(list(zip(xl,yl))) # plot empty due to (1,0)
+            doctest:warning
+            ...
             Graphics object consisting of 1 graphics primitive
 
         We remove `(1,0)` to fix this.::
--- a/sage/src/sage/plot/plot3d/list_plot3d.py
+++ b/sage/src/sage/plot/plot3d/list_plot3d.py
@@ -29,21 +29,20 @@
 
     OPTIONAL KEYWORDS:
 
-    - ``interpolation_type`` - 'linear', 'nn' (natural neighbor), 'spline'
+    - ``interpolation_type`` - 'linear', 'clough' (CloughTocher2D), 'spline'
 
       'linear' will perform linear interpolation
 
-      The option 'nn' An interpolation method for multivariate data in a 
-      Delaunay triangulation. The value for an interpolation point is 
-      estimated using weighted values of the closest surrounding points in 
-      the triangulation. These points, the natural neighbors, are the ones 
-      the interpolation point would connect to if inserted into the 
-      triangulation.
+      The option 'clough' will interpolate by using a piecewise cubic interpolating
+      Bezier polynomial on each triangle, using a Clough-Tocher scheme.
+      The interpolant is guaranteed to be continuously differentiable.
+      The gradients of the interpolant are chosen so that the curvature of the
+      interpolating surface is approximatively minimized.
 
       The option 'spline' interpolates using a bivariate B-spline.
 
       When v is a matrix the default is to use linear interpolation, when
-      v is a list of points the default is nearest neighbor.
+      v is a list of points the default is 'clough'.
 
     - ``degree`` - an integer between 1 and 5, controls the degree of spline
       used for spline interpolation. For data that is highly oscillatory
@@ -93,14 +92,14 @@
 
         sage: import warnings
         sage: warnings.simplefilter('ignore', UserWarning)
-        sage: list_plot3d(m, texture='yellow', interpolation_type='nn', frame_aspect_ratio=[1, 1, 1/3])
+        sage: list_plot3d(m, texture='yellow', interpolation_type='clough', frame_aspect_ratio=[1, 1, 1/3])
         Graphics3d Object
 
     We can make this look better by increasing the number of samples.
 
     ::
 
-        sage: list_plot3d(m, texture='yellow', interpolation_type='nn', frame_aspect_ratio=[1, 1, 1/3], num_points=40)
+        sage: list_plot3d(m, texture='yellow', interpolation_type='clough', frame_aspect_ratio=[1, 1, 1/3], num_points=40)
         Graphics3d Object
 
     Let's try a spline.
@@ -141,7 +140,7 @@
         sage: for i in range(-5, 5):
         ....:     for j in range(-5, 5):
         ....:         l.append((normalvariate(0, 1), normalvariate(0, 1), normalvariate(0, 1)))
-        sage: list_plot3d(l, interpolation_type='nn', texture='yellow', num_points=100)
+        sage: list_plot3d(l, interpolation_type='clough', texture='yellow', num_points=100)
         Graphics3d Object
 
     TESTS:
@@ -165,7 +164,7 @@
     different z coordinates, an exception will be raised::
 
         sage: pts =[(-4/5, -2/5, -2/5), (-4/5, -2/5, 2/5), (-4/5, 2/5, -2/5), (-4/5, 2/5, 2/5), (-2/5, -4/5, -2/5), (-2/5, -4/5, 2/5), (-2/5, -2/5, -4/5), (-2/5, -2/5, 4/5), (-2/5, 2/5, -4/5), (-2/5, 2/5, 4/5), (-2/5, 4/5, -2/5), (-2/5, 4/5, 2/5), (2/5, -4/5, -2/5), (2/5, -4/5, 2/5), (2/5, -2/5, -4/5), (2/5, -2/5, 4/5), (2/5, 2/5, -4/5), (2/5, 2/5, 4/5), (2/5, 4/5, -2/5), (2/5, 4/5, 2/5), (4/5, -2/5, -2/5), (4/5, -2/5, 2/5), (4/5, 2/5, -2/5), (4/5, 2/5, 2/5)]
-        sage: show(list_plot3d(pts, interpolation_type='nn'))
+        sage: show(list_plot3d(pts, interpolation_type='clough'))
         Traceback (most recent call last):
         ...
         ValueError: Two points with same x,y coordinates and different z coordinates were given. Interpolation cannot handle this.
@@ -173,7 +172,7 @@
     Additionally we need at least 3 points to do the interpolation::
 
         sage: mat = matrix(RDF, 1, 2, [3.2, 1.550])
-        sage: show(list_plot3d(mat, interpolation_type='nn'))
+        sage: show(list_plot3d(mat, interpolation_type='clough'))
         Traceback (most recent call last):
         ...
         ValueError: We need at least 3 points to perform the interpolation
@@ -320,18 +319,18 @@
 
     OPTIONAL KEYWORDS:
 
-    - ``interpolation_type`` - 'linear', 'nn' (natural neighbor), 'spline'
+    - ``interpolation_type`` - 'linear', 'clough' (CloughTocher2D), 'spline'
 
       'linear' will perform linear interpolation
 
-      The option 'nn' will interpolate by using natural neighbors. The 
-      value for an interpolation point is estimated using weighted values 
-      of the closest surrounding points in the triangulation.
+      The option 'clough' will interpolate by using a piecewise cubic interpolating
+      Bezier polynomial on each triangle, using a Clough-Tocher scheme.
+      The interpolant is guaranteed to be continuously differentiable.
 
       The option 'spline' interpolates using a bivariate B-spline.
 
       When v is a matrix the default is to use linear interpolation, when
-      v is a list of points the default is nearest neighbor.
+      v is a list of points the default is 'clough'.
 
     - ``degree`` - an integer between 1 and 5, controls the degree of spline
       used for spline interpolation. For data that is highly oscillatory
@@ -373,7 +372,7 @@
         sage: list_plot3d([(1, 2, 3), (0, 1, 3), (2, 1, 4), (1, 0, -2)], texture='yellow', num_points=50)
         Graphics3d Object
     """
-    from matplotlib import tri, delaunay
+    from matplotlib import tri
     import numpy
     import scipy
     from random import random
@@ -439,19 +438,15 @@
         G._set_extra_kwds(kwds)
         return G
 
-    if interpolation_type == 'nn'  or interpolation_type =='default':
+    if interpolation_type == 'clough'  or interpolation_type =='default':
 
-        T=delaunay.Triangulation(x,y)
-        f=T.nn_interpolator(z)
-        f.default_value=0.0
-        j=numpy.complex(0,1)
-        vals=f[ymin:ymax:j*num_points,xmin:xmax:j*num_points]
+        points=[[x[i],y[i]] for i in range(len(x))]
+        j = numpy.complex(0, 1)
+        f = interpolate.CloughTocher2DInterpolator(points,z)
         from .parametric_surface import ParametricSurface
-        def g(x,y):
-            i=round( (x-xmin)/(xmax-xmin)*(num_points-1) )
-            j=round( (y-ymin)/(ymax-ymin)*(num_points-1) )
-            z=vals[int(j),int(i)]
-            return (x,y,z)
+        def g(x, y):
+            z = f([x, y])
+            return (x, y, z)
         G = ParametricSurface(g, (list(numpy.r_[xmin:xmax:num_points*j]), list(numpy.r_[ymin:ymax:num_points*j])), texture=texture, **kwds)
         G._set_extra_kwds(kwds)
         return G
--- a/sage/src/sage/probability/probability_distribution.pyx
+++ b/sage/src/sage/probability/probability_distribution.pyx
@@ -118,7 +118,7 @@
             sage: P = [0.3, 0.4, 0.3]
             sage: X = GeneralDiscreteDistribution(P)
             sage: h, b = X.generate_histogram_data(bins = 10)
-            sage: h
+            sage: h  # rel tol 1e-08
             [1.6299999999999999,
              0.0,
              0.0,
--- a/sage/src/sage/plot/graphics.py
+++ b/sage/src/sage/plot/graphics.py
@@ -2457,6 +2457,7 @@
                    axes_pad=None, ticks_integer=None,
                    tick_formatter=None, ticks=None, title=None,
                    title_pos=None, base=None, scale=None,
+                   stylesheet='classic',
                    typeset='default'):
         r"""
         Return a matplotlib figure object representing the graphic
@@ -2526,6 +2527,11 @@
         if not isinstance(ticks, (list, tuple)):
             ticks = (ticks, None)
 
+        import matplotlib.pyplot as plt
+        if stylesheet not in plt.style.available:
+            stylesheet = 'classic'
+        plt.style.use(stylesheet)
+
         from sage.symbolic.ring import SR
         if not isinstance(tick_formatter, (list, tuple)):  # make sure both formatters typeset or both don't
             if tick_formatter == "latex" or tick_formatter in SR:
--- a/sage/src/doc/common/conf.py
+++ b/sage/src/doc/common/conf.py
@@ -26,6 +26,12 @@
     from sage.misc.temporary_file import tmp_filename
     import matplotlib.pyplot as plt
     if os.environ.get('SAGE_SKIP_PLOT_DIRECTIVE', 'no') != 'yes':
+        import matplotlib as mpl
+        mpl.rcParams['image.interpolation'] = 'bilinear'
+        mpl.rcParams['image.resample'] = False
+        mpl.rcParams['figure.figsize'] = [8.0, 6.0]
+        mpl.rcParams['figure.dpi'] = 80
+        mpl.rcParams['savefig.dpi'] = 100
         fn = tmp_filename(ext=".png")
         plot.plot().save(fn)
         img = mpimg.imread(fn)
--- a/sage/src/sage/doctest/control.py
+++ b/sage/src/sage/doctest/control.py
@@ -154,6 +154,11 @@
         """
         return not (self == other)
 
+# Make sure the agg backend is selected during doctesting
+# This needs to be done before any other matplotlib calls.
+import matplotlib
+matplotlib.use('agg')
+
 
 def skipdir(dirname):
     """
--- a/sage/src/sage/tests/startup.py
+++ b/sage/src/sage/tests/startup.py
@@ -3,10 +3,6 @@
 
 EXAMPLES::
 
-    sage: 'numpy' in sys.modules
-    False
-    sage: 'pyparsing' in sys.modules
-    False
     sage: 'sage.libs.gap.libgap' in sys.modules
     False
 
@@ -18,4 +14,17 @@
     sage: cmd = "from sage.all import *\nprint('IPython' in sys.modules)\n"
     sage: print(test_executable(["sage", "--python"], cmd)[0])  # long time
     False
+
+Check that numpy (:trac:`11714`) and pyparsing are not imported on startup
+as they increase the startup time. Since :trac:`23696` those are imported
+by the doctest framework via a matplotlib import. Again the simple would
+not work.
+
+    sage: from sage.tests.cmdline import test_executable
+    sage: cmd = "from sage.all import *\nprint('numpy' in sys.modules)\n"
+    sage: print(test_executable(["sage", "--python"], cmd)[0])  # long time
+    False
+    sage: cmd = "from sage.all import *\nprint('pyparsing' in sys.modules)\n"
+    sage: print(test_executable(["sage", "--python"], cmd)[0])  # long time
+    False
 """
